### DISCUSSION TITLE
Any plan for linkerd2-proxy to support WASM?

### DISCUSSION THREAD

**Discussion by ElvinEfendi** [linkerd/linkerd2#9723] [Ideas]
Envoy implements WASM support as per https://github.com/proxy-wasm/spec, would you consider similar feature in linkerd2-proxy?

There are scenarios where we need to do custom logic at the routing layer that might not make sense to be part of the proxy's core. For example, at my company, in certain cases, we look at pretty much every HTTP request attribute (host, path, header, cookie) and communicate with an external service to implement dynamic routing. That logic probably does not make sense to be part of linkerd2-proxy's core.

If there's an interest, we would also be interested in collaborating in implementing this.

**Comment by wmorgan** [comment]
Hi @ElvinEfendi! I think the answer is yes, at the macro level we're open to it, but there are a lot of details to sort out and a whole roadmap of other features clamoring for our attention. Certainly open to a discussion though!

**Reply by abergmeier** [reply]
I always wonder whether the more obvious/performant step wouldn't be to add support for plugins in linkerd2-proxy.
Since Rust ABI is not stable, the nearest to native performance would probably be to spawn an optional subprocess and send Protobuf encoded streams back and forth.
